daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/UserDefinedFunctionBase.scala

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dpath
+
+import org.apache.daffodil.exceptions.Assert
+import java.lang.reflect.Method
+import org.apache.daffodil.udf.UserDefinedFunction
+
+/**
+ * Both the evaluate method and the User Defined Function instance are passed in, as both are needed by the Method.invoke
+ * function.
+ */
+case class UserDefinedFunctionCall(recipes: List[CompiledDPath], userDefinedFunction: UserDefinedFunction, evaluateFxn: Method)
+  extends FNArgsList(recipes) {
+  override def computeValue(values: List[Any], dstate: DState) = {
+    val jValues = values.map { _.asInstanceOf[Object] }
+    val res = evaluateFxn.invoke(userDefinedFunction, jValues: _*)
```

Comment:
What happens if the UDF throws an Exception? Right not it will bubble up and probably cause some kind of error. Does it make sense here to catch generic exceptiosn and convert them to something else (maybe an SDE?)

Related, do we need a way for UDF's to throw an schema definition error or processing error? Or Both? Do we need special UDFExeception's that a UDF is expected to throw to signify SDE vs PE?

Reply:
That's a good point, I agree we ought to catch exceptions here (and in Expressions) and convert them to SDE. I'm not sure we should create a special UDFExceptions (and only catch that) as I'm more concerned about the Exceptions the UDF programmer doesn't account for.

At this point in the code, any exception caught would be a processing error versus a schema definition error, I think. Before this, we could argue for SDE or PE (which is what I think our logs have been representing).

Reply:
I think we would need API features to make it sensible for a UDF to signal an SDE based on the UDF author's logic. Since these are pure functions, ... other than the scenarios we already check for, I'm not sure what other situations warrant a SDE. We already do the type-checking. I would suggest we not hold up this change set for that, and if we find we need SDE we change later. 

I think a PE makes more sense, and anything thrown by a UDF should become a Processing error (parse or unparse error... don't forget about unparse!)  I think there is a constructor to create a PE from an Exception or Throwable directly. (If not there should be.)

Reply:
A potential case for an SDE: Imagine a UDF that accepts lat/long parameters and converts to some other geographic coordinate system. What's the right behavior if the parameters passed into this UDF are outside the range of a valid lat/long. It's not clear to me if this should be a PE or SDE, but looking at the spec, the dfdl: construction functions talk about schema defiition errors:

> If the string [parameter] begins with 'x', it is a schema definition error if a character appears other 0-9, a-f, A-F.

This to me feels similar to lat/long values out of range, make me think SDE is the right result in this particular case? I'm fine with postponing this kindof change though--SDE functionality can always be added later if deemed necessary.

Reply:
The DFDL constructors with the hex strings you mention, ... strings and numeric ranges really are pretty different. In this particular case, those functions have this artificial hex syntax feature intended to be used with literal constants e.g., dfdl:int("xA1B2C3D4"). They *could* be used with the argument coming direct from data, but how likely is it that data would use this prefix "x" convention to indicate a value expressed in hex? This is clearly for use with human authored constants, and that's why this is an SDE, not a PE.  Fact is, it *should* be a PE, and we should be able to rely on Daffodil to detect this PE at compile time for literal constants. I'm not sure we can do that right now. I think we eval expressions at compile time, but if they cause an error we just assume they can't be successfully evaluated at compile time, and so are not constants. So I think if we made these PE, then dfdl:int("foobar") would end up a PE at runtime, which we don't want. This is just a limitation of our expression compiler though. It *could* tell the difference between an error evaluating an expression due to constant args, vs. accessing the infoset (which doesn't exist at compile time). I just don't think it does right now.

Reply:
So for this, I think we're on the same page of it being a Processing Error. The question is if we should create a UserDefinedFunctionException class that UDFs can throw an error or if we should catch generic Exception and convert to Processing Error? No errors would be masked with the latter as it'd be converted to a PE.

Reply:
Expression evaluation is always surrounded by a try/catch which catches and converts things to PEs for you, so you don't need to reimplement that logic. 

What you do need to do is make sure the diagnostic is specific enough to enable the user to fix the problem, should it propagate all the way back up to top level and halt the parse. 

If you just allowed exceptions to propagate and didn't catch them at all, I think you'd get a PE that would say something like IllegalArgumentException, and give the file and line number of the whole expression that the UDF call appeared in. We want to be more specific than that. So I would try  catch, encapsulate as a PE and re-throw that.

Reply:
Our expression exception try-catch is defined in DPath.scala as "handleThrow()". By design, this block only captures exceptions which we expect to be triggerable by problems in the DPath expression itself; and (to the extent possible) not exceptions that are caused by genuine bugs in Daffodil. Since UDFs are implemented as arbitrary Java/Scala code, it is possible for them to throw any exception. For example, if a UDF throws a ClassCastException, we want to capture it and treat is as a some sort of user error [0]. However, if that same exception gets thrown from Daffodil code, it is a Daffodil Bug, and should get the "this is a bug in Daffodil" message.

I would say that we add a UserDefinedFunctionException say that UDFs are only allowed to throw UserDefinedFunctionException`s, which are interpreted as a correct result indicated that the DPath expression results in an error. Any exception thrown which is not a UserDefinedFunctionException is intereprated as the UDF implementation itself had an error, so we do not know if the DPath expression that was being computed should have succeeded.

If the exception being thrown is a UserDefinedFunctionException, I think we can put the responsibility on the UDF author to include sufficient information. For any other exception, we should provide backtrace of the code capturing the portion of the stack below UDF entry point.

Whatever we decide, I think error handling should be added to the design proposal. 

[0] originally I thought a standard Processing Error for which we can backtrack; but now that I write this out, I think an unsupported exception type should be fatal regardless of backtracking.

Reply:
So the current error handling is as follows:
For the case where an Exception is thrown during the evaluate call. (It gets captured by the Reflection class and comes out as a ReflectiveOperationException), we throw a processing error as our current expectation is not for the UDFs to be throwing errors. (I can see why we would want to change this, but more on that below)

All other exceptions thrown from UDFP load time till UDF initialization time, are caught and output as log errors/warnings. Because we can either not load any UDFs or drop the offender, and throw an SDE for unsupported function.

So the idea posited would mean we throw an SDE if we receive a UDFException from the UDF. As this would be them signalling that they received some data (or something) they considered invalid, so it wouldn't actually be a processing error, but instead a signal to Daffodil that something was wrong with the received parameters? That makes sense. I'll implement.

Re the backtrace, would it be preferable over something like the below?
[error] Error loading User Defined Function Providers: Error thrown in: org.sgoodudfs.example.StringFunctions.StringFunctionsProvider$CustomException: UDF Error!

Reply:
My proposal was that a UDFException be interperated as a processing error, and all other errors be interperated as some form of nonrecoverable error (either an RSDE, or some new class of error).

The idea being that if a UDF throws a UDFException, it is indicating that it considers the input data to be invalid. That is not nessaserilly a problem with the schema, just that the particular data being input is not valid. In such a case, it makes sense for Daffodil to backtrack and try a different branch. 

If the UDF throws any other kind of exception, it means that Daffodil encountered a bug in the UDF. Since we know this situation can only arise as a result of an error, there is no safe way to proceed from a correctness perspective (eg. the UDF did not signal that it was supposed to fail, so we do not know that backtracking is correct).

For the backtrace, I think we do want a backtrace showing all of the UDF portion of the stack. These functions could be highly non-trivial; possible being wrappers around a complex user implemented library. This situation will only arise when there is a programming error somewhere in the UDF and its dependency, so we should give the normal programming debugging output of a stacktrace.

Restricting the stacktrace to the UDF portion of the stack is more of a nice-to-have than a critical feature. Daffodil stack traces can get pretty long, and won't have any useful information for debugging the UDF. Plus, restricting the stacktrace will let me pretend that we are not just dynamically splicing user code into our running program.

Reply:
So let me summarize:
# user checks for an error (e.g., arg out of range) - author of UDF chooses that this should be an error that causes backtracking when parsing - we want a Processing Error. Author of UDF must be able to select this behavior, and distinguish from a fatal error, so we need UDFProcessingError(cause/message) class to throw.  I think this is a new class derived from ProcessingError that we need. 
# errors at setup time - SDE or added SDW that accompanies an SDE. These are all detected by daffodil, so nothing for users to do here. 
# unexpected error - user's code in the UDF just fails somehow - catch cause, and throw new UDFFatalException, which is an abort and captures the backtrace. This should throw to top level and gives us the option to say not "this is a bug in daffodil...." but to detect this class and issue a slightly different message. 

This last one is questionable. One can argue we should just not catch such exceptions, and they should fly out to top level and stop things. I think if the user does nothing to catch these, perhaps we should be doing nothing to catch these. 

I think UDF authors do need a positive way to cause a fatal exception that is assured NOT to be caught and cause a processing error. I would suggest:
will say (today) "This is a bug in daffodil...."
 
`class UDFFatalException(cause: Throwable) extends Abort(cause) `